gno.land/pkg/gnoland/app.go

```diff
@@ -196,9 +225,109 @@ func InitChainer(baseApp *sdk.BaseApp, acctKpr auth.AccountKeeperI, bankKpr bank
 	}
 }
 
-// XXX not used yet.
-func EndBlocker(vmk vm.VMKeeperI) func(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
-	return func(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
-		return abci.ResponseEndBlock{}
+// endBlockerApp is the app abstraction required by any EndBlocker
+type endBlockerApp interface {
+	// LastBlockHeight returns the latest app height
+	LastBlockHeight() int64
+
+	// Logger returns the logger reference
+	Logger() *slog.Logger
+}
+
+// EndBlocker defines the logic executed after every block.
+// Currently, it parses events that happened during execution to calculate
+// validator set changes
+func EndBlocker(
+	collector *collector[validatorUpdate],
+	vmKeeper vm.VMKeeperI,
+	app endBlockerApp,
+) func(
+	ctx sdk.Context,
+	req abci.RequestEndBlock,
+) abci.ResponseEndBlock {
+	return func(ctx sdk.Context, _ abci.RequestEndBlock) abci.ResponseEndBlock {
+		// Check if there was a valset change
+		if len(collector.getEvents()) == 0 {
+			// No valset updates
+			return abci.ResponseEndBlock{}
+		}
+
+		// Run the VM to get the updates from the chain
+		msg := vm.MsgCall{
```

Comment:
With this new approach, a read-only `Eval` should be sufficient. Did you try it?

Reply:
This is a great catch 💯 

I've tested it out, and it works great. I've swapped it to `QueryEval`:

[e628720](https://github.com/gnolang/gno/pull/2229/commits/e628720623a91c71a5a54f0a3468f3826731b985)